src/main/java/org/folio/inventory/support/KafkaConsumerVerticle.java

```diff
@@ -0,0 +1,144 @@
+import org.folio.inventory.storage.Storage;
+import org.folio.kafka.GlobalLoadSensor;
+import org.folio.kafka.KafkaConfig;
+import org.folio.kafka.KafkaConsumerWrapper;
+import org.folio.kafka.SubscriptionDefinition;
+import org.folio.okapi.common.GenericCompositeFuture;
+
+public abstract class KafkaConsumerVerticle extends AbstractVerticle {
+  private static final String LOAD_LIMIT_TEMPLATE = "inventory.kafka.%s.loadLimit";
+  private static final String LOAD_LIMIT_DEFAULT = "5";
+  private static final String MAX_DISTRIBUTION_NUMBER_TEMPLATE = "inventory.kafka.%s.maxDistributionNumber";
+  private static final String MAX_DISTRIBUTION_NUMBER_DEFAULT = "100";
+  private static final String CACHE_EXPIRATION_DEFAULT = "3600";
+  private static final String METADATA_EXPIRATION_TIME = "inventory.mapping-metadata-cache.expiration.time.seconds";
+  private final List<KafkaConsumerWrapper<String, String>> consumerWrappers = new ArrayList<>();
+  private KafkaConfig kafkaConfig;
+  private JsonObject config;
+  private HttpClient httpClient;
+  private Storage storage;
+  private MappingMetadataCache mappingMetadataCache;
+
+  @Override
+  public void stop(Promise<Void> stopPromise) {
+    List<Future<Void>> stopFutures = consumerWrappers.stream()
+      .map(KafkaConsumerWrapper::stop)
+      .toList();
+
+    GenericCompositeFuture.join(stopFutures)
+      .onComplete(ar -> stopPromise.complete());
+  }
+
+  protected abstract Logger getLogger();
+
+  protected KafkaConsumerWrapper<String, String> createConsumer(String eventType) {
+    var kafkaConsumerWrapper = KafkaConsumerWrapper.<String, String>builder()
+      .context(context)
+      .vertx(vertx)
+      .kafkaConfig(getKafkaConfig())
+      .loadLimit(getLoadLimit())
+      .globalLoadSensor(new GlobalLoadSensor())
+      .subscriptionDefinition(getSubscriptionDefinition(getKafkaConfig().getEnvId(), eventType))
+      .build();
+    consumerWrappers.add(kafkaConsumerWrapper);
+    return kafkaConsumerWrapper;
+  }
+
+  protected JsonObject getConfig() {
+    return ofNullable(config).orElseGet(() -> {
+      config = vertx.getOrCreateContext().config();
+      return config;
+    });
+  }
+
+  protected KafkaConfig getKafkaConfig() {
+    return ofNullable(kafkaConfig).orElseGet(() -> {
+      kafkaConfig = KafkaConfig.builder()
+        .envId(getConfig().getString(KAFKA_ENV))
+        .kafkaHost(getConfig().getString(KAFKA_HOST))
+        .kafkaPort(getConfig().getString(KAFKA_PORT))
+        .okapiUrl(getConfig().getString(OKAPI_URL))
+        .replicationFactor(parseInt(getConfig().getString(KAFKA_REPLICATION_FACTOR)))
+        .maxRequestSize(parseInt(getConfig().getString(KAFKA_MAX_REQUEST_SIZE)))
+        .build();
+      getLogger().info("kafkaConfig: {}", kafkaConfig);
+      return kafkaConfig;
+    });
+  }
+
+  protected HttpClient getHttpClient() {
+    return ofNullable(httpClient).orElseGet(() -> {
+      httpClient = vertx.createHttpClient();
+      return httpClient;
```

Comment:
is it lazy init in most of fields? 
could we avoid it and don't have runtime problems?

Reply:
discussed and agreed it's ok since it was before